src/logic/user-items-filter.ts

```diff
@@ -0,0 +1,150 @@
+  const ensureERC721 = (await config.getString('ENSURE_ERC_721')) === 'true'
+
+  /**
+   * Filter out wearables and emotes that are not owned by the user.
+   *
+   * @param {Avatar[]} avatars
+   * @param {string} owner
+   * @return {*}  {Promise<Avatar[]>}
+   */
+  async function filterNotOwnedItemsFromAvatars(avatars: Avatar[], owner: string): Promise<Avatar[]> {
+    const ownedWearables: OnChainWearable[] = await wearablesFetcher.fetchOwnedElements(owner)
+    const ownedEmotes: OnChainEmote[] = await emotesFetcher.fetchOwnedElements(owner)
+
+    // Filter out wearables and emotes that are not owned by the user.
+    const sanitizedAvatars: Avatar[] = avatars.map((avatar: Avatar) => {
+      const validatedWearables: string[] = parseValidWearablesAndFilterInvalidOnes(
+        avatar.avatar.wearables,
+        ownedWearables,
+        ensureERC721
+      )
+
+      const validatedEmotes = avatar.avatar.emotes
+        ?.map((emote: { urn: string; slot: number }) => {
+          if (!emote.urn.includes(':')) {
+            return emote
+          }
+
+          let urnToReturn: string = emote.urn
+
+          const { urn, tokenId } = splitUrnAndTokenId(emote.urn)
+
+          const matchingOwnedEmote = ownedEmotes.find(
+            (ownedEmote) =>
+              ownedEmote.urn === urn &&
+              (!tokenId || ownedEmote.individualData.find((itemData) => itemData.tokenId === tokenId))
+          )
+
+          if (!matchingOwnedEmote) {
+            return undefined
+          }
+
+          urnToReturn = ensureERC721
+            ? `${matchingOwnedEmote.urn}:${tokenId ? tokenId : matchingOwnedEmote.individualData[0].tokenId}`
+            : matchingOwnedEmote.urn
+
+          return { urn: urnToReturn, slot: emote.slot }
+        })
+        .filter((emote) => !!emote) as { urn: string; slot: number }[]
+
+      return { ...avatar, avatar: { ...avatar.avatar, wearables: validatedWearables, emotes: validatedEmotes } }
+    })
+
+    return sanitizedAvatars
+  }
+
+  /**
+   * Filter out outfits that contain wearables not owned by the user.
+   *
+   * @param {Outfits} outfits
+   * @param {string} owner
+   * @return {*}  {Promise<Outfits>}
+   */
+  async function filterOutfitsWithoutCompleteOwnership(outfits: Outfits, owner: string): Promise<Outfits> {
+    const ownedWearables: OnChainWearable[] = await wearablesFetcher.fetchOwnedElements(owner)
+
+    const outfitsWithOwnedWearables = outfits.outfits
+      .map((outfit) => {
+        if (!isValidOutfit(outfit.outfit, ownedWearables)) {
+          return undefined
+        }
+
+        const parsedOutfitsWearables = parseValidWearablesAndFilterInvalidOnes(
+          outfit.outfit.wearables,
+          ownedWearables,
+          ensureERC721
+        )
+
+        return { ...outfit, outfit: { ...outfit.outfit, wearables: parsedOutfitsWearables } }
+      })
+      .filter((outfit) => !!outfit) as { slot: number; outfit: Outfit }[]
```

Comment:
I would try to avoid the `as ..` it's very error-prone, I think in this case you could rewrite this as a for loop, I know it's a bit uglier but you would avoid typescript incorrectly infering types